### ISSUE TITLE
[Request] Dosbox API

### ISSUE THREAD

**Issue by Pantalaim0n** [joncampbell123/dosbox-x#775]
What I'd really like to see in DosBox is some sort of API to be able to receive data from and send commands to the DosBox process. For example, I love the way I'm able to control [MPC-HC](https://mpc-hc.org/) with the [many commands](https://github.com/Grokzen/mpcapi) available in its API.

Now, what I'm trying to achieve in DosBox for a while now is a way to automate things by emulating keypresses and/or mouse events, either by writing a TSR in DosBox itself, or by sending keypresses with the lowlevel Windows API [SendInput](https://msdn.microsoft.com/en-us/library/windows/desktop/ms646310.aspx) function, both not getting the results I would like.

It would be extremely helpful to be able to do this by 'talking' to DosBox in both directions, like sending said input events, or requesting data like current video resolution, mouse position, cursor position, etc.

Is this anything worth considering?

**Comment by joncampbell123** [comment]
Yes! The ability to automate DOSBox-X would be incredibly useful, including for automated testing.

There is a DOSBox-X integration device on I/O ports 28h-2Bh that you can talk to from a TSR within DOSBox-X itself. Currently, it's only used by the Windows 3.1 mouse driver in DOSLIB to allow VirtualBox style "pointer integration" so that mouse capture is not needed to drive the mouse pointer in Windows 3.1, but there are commands in place to allow stuffing mouse and keyboard input as well.

The interface is implemented in src/ints/bios.cpp.

**Comment by sikthehedgehog** [comment]
For the sake of completeness (since the "line 451" comment will rot whenever that file changes):
https://github.com/joncampbell123/dosbox-x/blob/eea85d18815d658e733865eff8da74f7176394a8/src/ints/bios.cpp#L451

I suppose what's needed then is a proper file documenting it so one doesn't have to wade through the source code to find the relevant parts.

**Comment by joncampbell123** [comment]
It's not a well enough developed "API" to write a document for, but there is code in DOSLIB to poke at it:

https://github.com/joncampbell123/doslib/tree/master/hw/dosboxid
